File: python/witan/workbook.py
from __future__ import annotations

import asyncio
import itertools
import os
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Any, Literal, cast

from ._async_process import AsyncStdioRPCProcess
from ._binary import get_binary_path
from ._process import StdioRPCProcess
from ._spreadsheet_base import _AsyncSpreadsheetSessionBase, _SpreadsheetSessionBase, drop_none
from .exceptions import WitanProcessError
from .types import (
    CellRef,
    DataTable,
    DataTableMutationResult,
    DataTableSpec,
    DataValidationInfo,
    DataValidationSpec,
    FormulaResult,
    ImageInfo,
    ImageSpec,
    ImageUpdate,
    JsonMapping,
    ListObject,
    ListObjectMutationResult,
    ListObjectSpec,
    ListObjectUpdate,
    SweepResult,
    WriteResult,
)


class _WorkbookBase:
    def __init__(
        self,
        path: str | os.PathLike[str],
        *,
        create: bool = False,
        locale: str | None = None,
        hint: str | None = None,
        stateless: bool | None = None,
        api_key: str | None = None,
        api_url: str | None = None,
        binary: str | os.PathLike[str] | None = None,
        env: Mapping[str, str] | None = None,
        request_timeout: float | None = None,
    ) -> None:
        self.path = Path(path)
        self.create = create
        self.locale = locale
        self.hint = hint
        self.stateless = stateless
        self.api_key = api_key
        self.api_url = api_url
        self.binary = Path(binary) if binary is not None else None
        self.env = dict(env or {})
        self.request_timeout = request_timeout
        self._ids = itertools.count(1)

    def _next_id(self) -> str:
        return str(next(self._ids))

    def _argv(self) -> list[str]:
        binary = str(self.binary) if self.binary is not None else get_binary_path()
        argv = [binary]
        if self.api_key is not None:
            argv.extend(["--api-key", self.api_key])
        if self.api_url is not None:
            argv.extend(["--api-url", self.api_url])
        if self.stateless is True:
            argv.append("--stateless")
        argv.extend(["xlsx", "rpc", str(self.path)])
        if self.create:
            argv.append("--create")
        if self.hint is not None:
            argv.extend(["--hint", self.hint])
        if self.locale is not None:
            argv.extend(["--locale", self.locale])
        return argv


class Workbook(_WorkbookBase, _SpreadsheetSessionBase):
    """Synchronous Witan workbook session backed by `witan xlsx rpc`."""

    def __init__(
        self,
        path: str | os.PathLike[str],
        *,
        create: bool = False,
        locale: str | None = None,
        hint: str | None = None,
        stateless: bool | None = None,
        api_key: str | None = None,
        api_url: str | None = None,
        binary: str | os.PathLike[str] | None = None,
        env: Mapping[str, str] | None = None,
        request_timeout: float | None = None,
    ) -> None:
        _WorkbookBase.__init__(
            self,
            path,
            create=create,
            locale=locale,
            hint=hint,
            stateless=stateless,
            api_key=api_key,
            api_url=api_url,
            binary=binary,
            env=env,
            request_timeout=request_timeout,
        )
        _SpreadsheetSessionBase.__init__(self)
        self._process = StdioRPCProcess(self._argv(), env=self.env, request_timeout=request_timeout)

    def __enter__(self) -> "Workbook":
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        self.close()

    def _request(self, method: str, op: str, args: Mapping[str, Any] | None = None) -> Any:
        return self._process.request(
            request_id=self._next_id(),
            op=op,
            args=args or {},
            method=method,
        )

    def close(self) -> None:
        self._process.close()

    def save(self) -> bool:
        return cast(bool, self._request("save", "save", {}))

    def get_data_validations(self, *, sheet: str | None = None, address: str | None = None) -> list[DataValidationInfo]:
        result = cast(Mapping[str, Any], self._request("get_data_validations", "getDataValidations", drop_none({"sheet": sheet, "address": address})))
        return cast(list[DataValidationInfo], result.get("rules", []))

    def set_data_validations(self, sheet_name: str, rules: Sequence[DataValidationSpec], *, clear: bool | None = None) -> None:
        self._request("set_data_validations", "setDataValidations", drop_none({"sheet": sheet_name, "rules": list(rules), "clear": clear}))

    def remove_data_validations(self, sheet_name: str, *, indices: Sequence[int] | None = None, address: str | None = None) -> None:
        self._request("remove_data_validations", "removeDataValidations", drop_none({"sheet": sheet_name, "indices": list(indices) if indices is not None else None, "address": address}))

    # The ops below are not supported by the Google Sheets backend
    # (NOT_IMPLEMENTED), so they live on Workbook rather than the shared base.

    def get_list_object(self, name: str) -> ListObject:
        return cast(ListObject, self._request("get_list_object", "getListObject", {"name": name}))

    def add_list_object(self, sheet_name: str, list_object: ListObjectSpec) -> ListObjectMutationResult:
        return cast(ListObjectMutationResult, self._request("add_list_object", "addListObject", {"sheet": sheet_name, "listObject": list_object}))

    def set_list_object(self, name: str, list_object: ListObjectUpdate) -> ListObjectMutationResult:
        return cast(ListObjectMutationResult, self._request("set_list_object", "setListObject", {"name": name, "listObject": list_object}))

    def delete_list_object(self, name: str) -> WriteResult:
        return cast(WriteResult, self._request("delete_list_object", "deleteListObject", {"name": name}))

    def get_data_table(self, address: str) -> DataTable:
        return cast(DataTable, self._request("get_data_table", "getDataTable", {"address": address}))

    def add_data_table(self, sheet_name: str, data_table: DataTableSpec) -> DataTableMutationResult:
        return cast(DataTableMutationResult, self._request("add_data_table", "addDataTable", {"sheet": sheet_name, "dataTable": data_table}))

    def delete_data_table(self, address: str) -> WriteResult:
        return cast(WriteResult, self._request("delete_data_table", "deleteDataTable", {"address": address}))

    def list_images(self, *, sheet: str | None = None) -> list[ImageInfo]:
        result = cast(Mapping[str, Any], self._request("list_images", "listImages", drop_none({"sheet": sheet})))
        return cast(list[ImageInfo], result.get("images", []))

    def get_image(self, sheet: str, *, name: str | None = None, id: int | None = None) -> ImageInfo:
        result = cast(Mapping[str, Any], self._request("get_image", "getImage", drop_none({"sheet": sheet, "name": name, "id": id})))
        return cast(ImageInfo, result.get("image", {}))

    def add_image(self, sheet: str, image: ImageSpec) -> ImageInfo:
        result = cast(Mapping[str, Any], self._request("add_image", "addImage", {"sheet": sheet, "image": image}))
        return cast(ImageInfo, result.get("image", {}))

    def set_image(self, sheet: str, image: ImageUpdate, *, name: str | None = None, id: int | None = None) -> ImageInfo:
        result = cast(Mapping[str, Any], self._request("set_image", "setImage", drop_none({"sheet": sheet, "name": name, "id": id, "image": image})))
        return cast(ImageInfo, result.get("image", {}))

    def delete_image(self, sheet: str, *, name: str | None = None, id: int | None = None) -> None:
        self._request("delete_image", "deleteImage", drop_none({"sheet": sheet, "name": name, "id": id}))

    def sweep_inputs(self, inputs: Sequence[JsonMapping], outputs: Sequence[str | CellRef], *, mode: Literal["cartesian", "parallel"] | None = None, include_stats: bool | None = None) -> SweepResult:
        return cast(SweepResult, self._request("sweep_inputs", "sweepInputs", drop_none({"inputs": list(inputs), "outputs": list(outputs), "mode": mode, "includeStats": include_stats})))

    def scenarios(self, inputs: Sequence[JsonMapping], outputs: Sequence[str | CellRef], *, mode: Literal["cartesian", "parallel"] | None = None, include_stats: bool | None = None) -> SweepResult:
        return self.sweep_inputs(inputs, outputs, mode=mode, include_stats=include_stats)

    def evaluate_formulas(self, sheet: str, formulas: Sequence[str]) -> list[FormulaResult]:
        return cast(list[FormulaResult], self._request("evaluate_formulas", "evaluateFormulas", {"sheet": sheet, "formulas": list(formulas)}))

    def evaluate_formula(self, sheet: str, formula: str) -> FormulaResult:
        return self.evaluate_formulas(sheet, [formula])[0]


class AsyncWorkbook(_WorkbookBase, _AsyncSpreadsheetSessionBase):
    """Asynchronous Witan workbook session backed by `witan xlsx rpc`."""

    def __init__(
        self,
        path: str | os.PathLike[str],
        *,
        create: bool = False,
        locale: str | None = None,
        hint: str | None = None,
        stateless: bool | None = None,
        api_key: str | None = None,
        api_url: str | None = None,
        binary: str | os.PathLike[str] | None = None,
        env: Mapping[str, str] | None = None,
        request_timeout: float | None = None,
    ) -> None:
        _WorkbookBase.__init__(
            self,
            path,
            create=create,
            locale=locale,
            hint=hint,
            stateless=stateless,
            api_key=api_key,
            api_url=api_url,
            binary=binary,
            env=env,
            request_timeout=request_timeout,
        )
        _AsyncSpreadsheetSessionBase.__init__(self)
        self._process: AsyncStdioRPCProcess | None = None
        self._start_lock = asyncio.Lock()
        self._closed = False

    async def __aenter__(self) -> "AsyncWorkbook":
        await self._ensure_process()
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        await self.close()

    async def _ensure_process(self) -> AsyncStdioRPCProcess:
        if self._closed:
            raise WitanProcessError("witan subprocess is closed")
        process = self._process
        if process is not None:
            if process.closed:
                raise WitanProcessError("witan subprocess is closed")
            return process

        async with self._start_lock:
            if self._closed:
                raise WitanProcessError("witan subprocess is closed")
            process = self._process
            if process is not None:
                if process.closed:
                    raise WitanProcessError("witan subprocess is closed")
                return process
            self._process = await AsyncStdioRPCProcess.create(self._argv(), env=self.env, request_timeout=self.request_timeout)
            return self._process

    async def _request(self, method: str, op: str, args: Mapping[str, Any] | None = None) -> Any:
        process = await self._ensure_process()
        return await process.request(request_id=self._next_id(), op=op, args=args or {}, method=method)

    async def close(self) -> None:
        async with self._start_lock:
            if self._closed:
                return
            self._closed = True
            if self._process is not None:
                await self._process.close()

    async def save(self) -> bool:
        return cast(bool, await self._request("save", "save", {}))

    async def get_data_validations(self, *, sheet: str | None = None, address: str | None = None) -> list[DataValidationInfo]:
        result = cast(Mapping[str, Any], await self._request("get_data_validations", "getDataValidations", drop_none({"sheet": sheet, "address": address})))
        return cast(list[DataValidationInfo], result.get("rules", []))

    async def set_data_validations(self, sheet_name: str, rules: Sequence[DataValidationSpec], *, clear: bool | None = None) -> None:
        await self._request("set_data_validations", "setDataValidations", drop_none({"sheet": sheet_name, "rules": list(rules), "clear": clear}))

    async def remove_data_validations(self, sheet_name: str, *, indices: Sequence[int] | None = None, address: str | None = None) -> None:
        await self._request("remove_data_validations", "removeDataValidations", drop_none({"sheet": sheet_name, "indices": list(indices) if indices is not None else None, "address": address}))

    # The ops below are not supported by the Google Sheets backend
    # (NOT_IMPLEMENTED), so they live on AsyncWorkbook rather than the shared base.

    async def get_list_object(self, name: str) -> ListObject:
        return cast(ListObject, await self._request("get_list_object", "getListObject", {"name": name}))

    async def add_list_object(self, sheet_name: str, list_object: ListObjectSpec) -> ListObjectMutationResult:
        return cast(ListObjectMutationResult, await self._request("add_list_object", "addListObject", {"sheet": sheet_name, "listObject": list_object}))

    async def set_list_object(self, name: str, list_object: ListObjectUpdate) -> ListObjectMutationResult:
        return cast(ListObjectMutationResult, await self._request("set_list_object", "setListObject", {"name": name, "listObject": list_object}))

    async def delete_list_object(self, name: str) -> WriteResult:
        return cast(WriteResult, await self._request("delete_list_object", "deleteListObject", {"name": name}))

    async def get_data_table(self, address: str) -> DataTable:
        return cast(DataTable, await self._request("get_data_table", "getDataTable", {"address": address}))

    async def add_data_table(self, sheet_name: str, data_table: DataTableSpec) -> DataTableMutationResult:
        return cast(DataTableMutationResult, await self._request("add_data_table", "addDataTable", {"sheet": sheet_name, "dataTable": data_table}))

    async def delete_data_table(self, address: str) -> WriteResult:
        return cast(WriteResult, await self._request("delete_data_table", "deleteDataTable", {"address": address}))

    async def list_images(self, *, sheet: str | None = None) -> list[ImageInfo]:
        result = cast(Mapping[str, Any], await self._request("list_images", "listImages", drop_none({"sheet": sheet})))
        return cast(list[ImageInfo], result.get("images", []))

    async def get_image(self, sheet: str, *, name: str | None = None, id: int | None = None) -> ImageInfo:
        result = cast(Mapping[str, Any], await self._request("get_image", "getImage", drop_none({"sheet": sheet, "name": name, "id": id})))
        return cast(ImageInfo, result.get("image", {}))

    async def add_image(self, sheet: str, image: ImageSpec) -> ImageInfo:
        result = cast(Mapping[str, Any], await self._request("add_image", "addImage", {"sheet": sheet, "image": image}))
        return cast(ImageInfo, result.get("image", {}))

    async def set_image(self, sheet: str, image: ImageUpdate, *, name: str | None = None, id: int | None = None) -> ImageInfo:
        result = cast(Mapping[str, Any], await self._request("set_image", "setImage", drop_none({"sheet": sheet, "name": name, "id": id, "image": image})))
        return cast(ImageInfo, result.get("image", {}))

    async def delete_image(self, sheet: str, *, name: str | None = None, id: int | None = None) -> None:
        await self._request("delete_image", "deleteImage", drop_none({"sheet": sheet, "name": name, "id": id}))

    async def sweep_inputs(self, inputs: Sequence[JsonMapping], outputs: Sequence[str | CellRef], *, mode: Literal["cartesian", "parallel"] | None = None, include_stats: bool | None = None) -> SweepResult:
        return cast(SweepResult, await self._request("sweep_inputs", "sweepInputs", drop_none({"inputs": list(inputs), "outputs": list(outputs), "mode": mode, "includeStats": include_stats})))

    async def scenarios(self, inputs: Sequence[JsonMapping], outputs: Sequence[str | CellRef], *, mode: Literal["cartesian", "parallel"] | None = None, include_stats: bool | None = None) -> SweepResult:
        return await self.sweep_inputs(inputs, outputs, mode=mode, include_stats=include_stats)

    async def evaluate_formulas(self, sheet: str, formulas: Sequence[str]) -> list[FormulaResult]:
        return cast(list[FormulaResult], await self._request("evaluate_formulas", "evaluateFormulas", {"sheet": sheet, "formulas": list(formulas)}))

    async def evaluate_formula(self, sheet: str, formula: str) -> FormulaResult:
        return (await self.evaluate_formulas(sheet, [formula]))[0]

File: node/src/workbook.ts
import { getBinaryPath } from './binary.js';
import { WitanProcessError } from './errors.js';
import { dropUndefined } from './helpers.js';
import { StdioRPCProcess } from './process.js';
import { SpreadsheetSessionBase } from './spreadsheet-base.js';
import type {
  DataTable,
  DataTableMutationResult,
  DataTableSpec,
  DataValidationInfo,
  DataValidationSpec,
  FormulaResult,
  ImageInfo,
  ImageSelector,
  ImageSpec,
  ImageUpdate,
  ListObject,
  ListObjectMutationResult,
  ListObjectSpec,
  ListObjectUpdate,
  SweepInput,
  SweepMode,
  SweepResult,
  WriteResult,
} from './types.js';

/**
 * Options for opening a Workbook.
 */
export interface WorkbookOptions {
  /** Create the file if it doesn't exist */
  create?: boolean;
  /** Locale for number/date formatting */
  locale?: string;
  /** Hint for file format detection */
  hint?: string;
  /** Enable stateless mode */
  stateless?: boolean;
  /** API key for authentication */
  apiKey?: string;
  /** API URL override */
  apiUrl?: string;
  /** Custom binary path (overrides auto-detection) */
  binary?: string;
  /** Additional environment variables for the subprocess */
  env?: Record<string, string>;
  /** Request timeout in milliseconds (default: 90000). */
  requestTimeoutMs?: number;
}

/**
 * Async Workbook session backed by `witan xlsx rpc`.
 *
 * Use the static `Workbook.open()` factory to create instances.
 * Implements `AsyncDisposable` for use with `await using`.
 *
 * @example
 * ```typescript
 * // With await using (recommended)
 * {
 *   await using wb = await Workbook.open('report.xlsx');
 *   const sheets = await wb.listSheets();
 *   await wb.save();
 * } // Automatically closed
 *
 * // With try/finally
 * const wb = await Workbook.open('report.xlsx');
 * try {
 *   const sheets = await wb.listSheets();
 * } finally {
 *   await wb.close();
 * }
 * ```
 */
export class Workbook extends SpreadsheetSessionBase implements AsyncDisposable {
  private process: StdioRPCProcess;
  private requestId = 0;
  private closed = false;

  private constructor(process: StdioRPCProcess) {
    super();
    this.process = process;
  }

  /**
   * Open a workbook and start the RPC subprocess.
   *
   * @param path - Path to the Excel file
   * @param options - Configuration options
   * @returns A ready-to-use Workbook instance
   * @throws {WitanProcessError} If the subprocess fails to start
   */
  static async open(path: string, options: WorkbookOptions = {}): Promise<Workbook> {
    const argv = Workbook.buildArgv(path, options);
    const process = new StdioRPCProcess(argv, {
      env: options.env,
      timeoutMs: options.requestTimeoutMs,
    });

    try {
      await process.waitReady();
    } catch (err) {
      process.terminate();
      throw err;
    }

    return new Workbook(process);
  }

  private static buildArgv(path: string, options: WorkbookOptions): string[] {
    const binary = options.binary ?? getBinaryPath();
    const argv = [binary];

    if (options.apiKey !== undefined) {
      argv.push('--api-key', options.apiKey);
    }
    if (options.apiUrl !== undefined) {
      argv.push('--api-url', options.apiUrl);
    }
    if (options.stateless === true) {
      argv.push('--stateless');
    }

    argv.push('xlsx', 'rpc', path);

    if (options.create === true) {
      argv.push('--create');
    }
    if (options.hint !== undefined) {
      argv.push('--hint', options.hint);
    }
    if (options.locale !== undefined) {
      argv.push('--locale', options.locale);
    }

    return argv;
  }

  private nextId(): string {
    return String(++this.requestId);
  }

  protected override async request(
    method: string,
    op: string,
    args: Record<string, unknown> = {}
  ): Promise<unknown> {
    if (this.closed) {
      throw new WitanProcessError('Workbook is closed');
    }
    return this.process.request(method, op, args, this.nextId());
  }

  /**
   * Close the workbook and terminate the subprocess.
   * Safe to call multiple times.
   */
  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    await this.process.close();
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  /** Whether the workbook has been closed. */
  get isClosed(): boolean {
    return this.closed;
  }

  /**
   * Save the workbook to disk.
   * @returns true if the save was successful
   */
  async save(): Promise<boolean> {
    return (await this.request('save', 'save', {})) as boolean;
  }

  /**
   * Get data validation rules.
   *
   * @param options - Optional sheet and address filters
   * @returns Array of data validation rules
   */
  async getDataValidations(options: { sheet?: string; address?: string } = {}): Promise<DataValidationInfo[]> {
    const result = (await this.request(
      'getDataValidations',
      'getDataValidations',
      dropUndefined({ sheet: options.sheet, address: options.address })
    )) as { rules?: DataValidationInfo[] };
    return result.rules ?? [];
  }

  /**
   * Add data validation rules to a sheet.
   *
   * @param sheet - Sheet name
   * @param rules - Rules to add
   * @param options - Options including whether to clear existing rules
   */
  async setDataValidations(
    sheet: string,
    rules: DataValidationSpec[],
    options: { clear?: boolean } = {}
  ): Promise<void> {
    await this.request(
      'setDataValidations',
      'setDataValidations',
      dropUndefined({ sheet, rules, clear: options.clear })
    );
  }

  /**
   * Remove data validation rules by index or by range.
   *
   * @param sheet - Sheet name
   * @param target - Either rule indices or an address to clear
   */
  async removeDataValidations(sheet: string, target: { indices: number[] } | { address: string }): Promise<void> {
    await this.request('removeDataValidations', 'removeDataValidations', {
      sheet,
      ...target,
    });
  }

  // ============================================================================
  // The ops below are not supported by the Google Sheets backend
  // (NOT_IMPLEMENTED), so they live on Workbook rather than the shared base.
  // ============================================================================

  /** Get a list object (table) by name. */
  async getListObject(name: string): Promise<ListObject> {
    return (await this.request('getListObject', 'getListObject', { name })) as ListObject;
  }

  /** Add a list object (table) to a sheet. */
  async addListObject(sheet: string, listObject: ListObjectSpec): Promise<ListObjectMutationResult> {
    return (await this.request('addListObject', 'addListObject', {
      sheet,
      listObject,
    })) as ListObjectMutationResult;
  }

  /** Update a list object (table). */
  async setListObject(name: string, listObject: ListObjectUpdate): Promise<ListObjectMutationResult> {
    return (await this.request('setListObject', 'setListObject', {
      name,
      listObject,
    })) as ListObjectMutationResult;
  }

  /** Delete a list object (table). */
  async deleteListObject(name: string): Promise<WriteResult> {
    return (await this.request('deleteListObject', 'deleteListObject', { name })) as WriteResult;
  }

  /** Get a data table by address. */
  async getDataTable(address: string): Promise<DataTable> {
    return (await this.request('getDataTable', 'getDataTable', { address })) as DataTable;
  }

  /** Add a data table to a sheet. */
  async addDataTable(sheet: string, dataTable: DataTableSpec): Promise<DataTableMutationResult> {
    return (await this.request('addDataTable', 'addDataTable', {
      sheet,
      dataTable,
    })) as DataTableMutationResult;
  }

  /** Delete a data table. */
  async deleteDataTable(address: string): Promise<WriteResult> {
    return (await this.request('deleteDataTable', 'deleteDataTable', { address })) as WriteResult;
  }

  /** List worksheet images in the workbook. */
  async listImages(options: { sheet?: string } = {}): Promise<ImageInfo[]> {
    const result = (await this.request(
      'listImages',
      'listImages',
      dropUndefined({ sheet: options.sheet })
    )) as { images?: ImageInfo[] };
    return result.images ?? [];
  }

  /** Get worksheet image metadata by name or id. */
  async getImage(sheet: string, selector: ImageSelector): Promise<ImageInfo> {
    const result = (await this.request(
      'getImage',
      'getImage',
      dropUndefined({ sheet, name: selector.name, id: selector.id })
    )) as { image?: ImageInfo };
    return result.image ?? ({} as ImageInfo);
  }

  /** Add a PNG or JPEG image to a sheet. */
  async addImage(sheet: string, image: ImageSpec): Promise<ImageInfo> {
    const result = (await this.request('addImage', 'addImage', { sheet, image })) as {
      image?: ImageInfo;
    };
    return result.image ?? ({} as ImageInfo);
  }

  /** Update worksheet image metadata, placement, or source bytes. */
  async setImage(sheet: string, selector: ImageSelector, image: ImageUpdate): Promise<ImageInfo> {
    const result = (await this.request(
      'setImage',
      'setImage',
      dropUndefined({ sheet, name: selector.name, id: selector.id, image })
    )) as { image?: ImageInfo };
    return result.image ?? ({} as ImageInfo);
  }

  /** Delete a worksheet image by name or id. */
  async deleteImage(sheet: string, selector: ImageSelector): Promise<void> {
    await this.request(
      'deleteImage',
      'deleteImage',
      dropUndefined({ sheet, name: selector.name, id: selector.id })
    );
  }

  /** Run a sweep over input combinations and capture outputs. */
  async sweepInputs(
    inputs: SweepInput[],
    outputs: string[],
    options: { mode?: SweepMode; includeStats?: boolean } = {}
  ): Promise<SweepResult> {
    return (await this.request(
      'sweepInputs',
      'sweepInputs',
      dropUndefined({
        inputs,
        outputs,
        mode: options.mode,
        includeStats: options.includeStats,
      })
    )) as SweepResult;
  }

  /** Alias for sweepInputs. */
  async scenarios(
    inputs: SweepInput[],
    outputs: string[],
    options: { mode?: SweepMode; includeStats?: boolean } = {}
  ): Promise<SweepResult> {
    return this.sweepInputs(inputs, outputs, options);
  }

  /** Evaluate multiple formulas in a sheet context. */
  async evaluateFormulas(sheet: string, formulas: string[]): Promise<FormulaResult[]> {
    return (await this.request('evaluateFormulas', 'evaluateFormulas', {
      sheet,
      formulas,
    })) as FormulaResult[];
  }

  /** Evaluate a single formula in a sheet context. */
  async evaluateFormula(sheet: string, formula: string): Promise<FormulaResult> {
    const results = await this.evaluateFormulas(sheet, [formula]);
    return results[0]!;
  }
}
